crates/proof-of-sql/src/sql/proof_gadgets/membership_check_test.rs

```diff
@@ -0,0 +1,539 @@
+//! This module contains the implementation of the `MembershipCheckTestPlan` struct. This struct
+//! is used to check whether the membership check gadgets work correctly.
+use super::membership_check::{
+    final_round_evaluate_membership_check, first_round_evaluate_membership_check,
+    verify_membership_check,
+};
+use crate::{
+    base::{
+        database::{
+            ColumnField, ColumnRef, OwnedTable, Table, TableEvaluation, TableOptions, TableRef,
+        },
+        map::{indexset, IndexMap, IndexSet},
+        proof::ProofError,
+        scalar::Scalar,
+    },
+    sql::proof::{
+        FinalRoundBuilder, FirstRoundBuilder, ProofPlan, ProverEvaluate, VerificationBuilder,
+    },
+};
+use bumpalo::{
+    collections::{vec::Vec as BumpVec, CollectIn},
+    Bump,
+};
+use serde::Serialize;
+
+#[derive(Debug, Serialize)]
+pub struct MembershipCheckTestPlan {
+    pub source_table: TableRef,
+    pub candidate_table: TableRef,
+    pub source_columns: Vec<ColumnRef>,
+    pub candidate_columns: Vec<ColumnRef>,
+    pub proposed_multiplicities: Vec<usize>,
+}
+
+impl ProverEvaluate for MembershipCheckTestPlan {
+    #[doc = "Evaluate the query, modify `FirstRoundBuilder` and return the result."]
+    fn first_round_evaluate<'a, S: Scalar>(
+        &self,
+        builder: &mut FirstRoundBuilder<'a, S>,
+        alloc: &'a Bump,
+        table_map: &IndexMap<TableRef, Table<'a, S>>,
+    ) -> Table<'a, S> {
+        // Check that the source and candidate columns have the same amount of columns
+        assert_eq!(
+            self.source_columns.len(),
+            self.candidate_columns.len(),
+            "Source and candidate columns must have the same length"
+        );
```

Comment:
Can you put this check inside `first_round_evaluate_membership_check` and `final_round_evaluate_membership_check` instead?

Reply:
Actually it is not doable within `first_round_evaluate_membership_check` but will be done in `final_round_evaluate_membership_check`.